File: 105-construct-binary-tree-from-preorder-and-inorder-traversal/105-construct-binary-tree-from-preorder-and-inorder-traversal.scala
object Solution {
  def buildTree(preorder: Array[Int], inorder: Array[Int]): TreeNode = {
    var map: Map[Int, Int] = Map()

    inorder.zipWithIndex.foreach { case (v, i) => map += (v -> i) }

    helper(0, preorder.length - 1, 0, inorder.length - 1)(
      preorder,
      inorder,
      map
    )
  }

  def helper(ps: Int, pe: Int, is: Int, ie: Int)(
      preorder: Array[Int],
      inorder: Array[Int],
      map: Map[Int, Int]
  ): TreeNode = {
    if (ps <= pe) {
      val v = preorder(ps)
      val res = new TreeNode(v)
      val index = map.get(v).get
      val lsize = index - 1 - is + 1
      val rsize = ie - index

      res.left =
        helper(ps + 1, ps + lsize, is, index - 1)(preorder, inorder, map)
      res.right =
        helper(ps + lsize + 1, pe, index + 1, ie)(preorder, inorder, map)

      res
    } else null
  }
}

File: 105-construct-binary-tree-from-preorder-and-inorder-traversal/105-construct-binary-tree-from-preorder-and-inorder-traversal.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public TreeNode buildTree(int[] preorder, int[] inorder) {
        
        Map<Integer, Integer> idx = new HashMap<>();
        
        for(int i = 0; i < inorder.length; i++) {
            idx.put(inorder[i], i);
        }
        
        return helper(preorder, 0, preorder.length - 1, inorder, 0, inorder.length - 1, idx);
    }
    
    TreeNode helper(int[] preorder, int sp, int ep, int[] inorder, int si, int ei, Map<Integer, Integer> idx) {
        if (sp > ep) {
            return null;
        }
        
        TreeNode res = new TreeNode(preorder[sp]);
        
        int i = idx.get(preorder[sp]);
        int lsz = i - si;
        int rsz = ei - i;
        
        res.left = helper(preorder, sp + 1, sp + lsz, inorder, si, i - 1, idx);
        res.right = helper(preorder, sp + lsz + 1, ep, inorder, i + 1, ei, idx);
        
        return res;
    }
}